In [1]:
# execute to import notebook styling for tables and width etc.
from IPython.core.display import HTML
import urllib.request
response = urllib.request.urlopen('https://raw.githubusercontent.com/DataScienceUWL/DS775v2/master/ds755.css')
HTML(response.read().decode("utf-8"));

<font size=18>Lesson 10 - Self-Assessment Solutions</font>

# *Self-Assessment: Load and Display - Solution*

There's nothing too new here. You've done this kind of work before. What's more important here than the code is making sure you take a minute or two to understand the data you're pulling in. What columns do you have available to you? Which columns contain simple values and which columns contain lists. Think about how you could or couldn't use this data to make recommendations.

In [2]:
import pandas as pd
import numpy as np

ted = pd.read_csv('./data/ted-talks/ted_main.csv')
ted.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  published_date  \
0      Ken Robinson: Do schools kill creativity?            1      1151367060   
1           Al Gore: Averting the climate crisis            1      1151367060   
2                  David Pogue: Simplicity sells            1      1151367060   
3             Majora Carter: Greening the ghetto            1      1151367060   
4  Hans Rosling: The best stats you've ever seen            1      1151440680   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869

# *Self-Assessment: Pandas - Solution*
Remember that shape gives you the number of rows first, followed by the number of columns.

In [3]:
ted.shape

(2550, 17)

There are 2550 TED talks in this data frame.

# *Self-Assessment: Prerequisites - Solution*

Remember that when you're calculating the quantile for some piece of data, you'll get different results if you calculate it before or after you do your other subsetting. First, let's calculate the views quantile before we figure the rest of our prerequisites.

In [4]:
#Calculate the number of views for the 10th percentile - calculated from the whole dataframe
m = ted['views'].quantile(0.10)

#Only consider talks of at least 5 minutes
q_talks = ted[(ted['duration'] >= 300)]

#Only consider talks with one speaker
q_talks = q_talks[q_talks['num_speaker']==1]

#Only consider talks in the top 90%
q_talks = q_talks[q_talks['views'] >= m]

#Inspect the number of talks that made the cut
q_talks.shape

(2107, 17)

Let's compare that with calculating the quantile after we subset.

In [5]:
#Only consider talks of at least 5 minutes
q_talks2 = ted[(ted['duration'] >= 300)]

#Only consider talks with one speaker
q_talks2 = q_talks2[q_talks2['num_speaker']==1]

#Calculate the number of views for the 10th percentile - calculated from the whole dataframe
m2 = q_talks2['views'].quantile(0.10)

#Only consider talks in the top 90%
q_talks2 = q_talks2[q_talks2['views'] >= m2]

#Inspect the number of talks that made the cut
q_talks2.shape

(2093, 17)

There is no universally "right" answer as to whether you should calculate the quantile before or after you've narrowed the initial dataset. It depends on what you're trying to accomplish. If you want the most viewed talks *that meet your criteria* you'd calculate it after you've subsetted. If you want the most viewed talks *overall* you'd calculate it before you've subsetted.

For our homework, we'll either tell you when to subset a dataframe or ask you to make the decision and give a justification for your decision.

# *Self-Assessment: Compute a Metric, Sort and Print - Solution*

Note that here we are computing our metric on our narrowed data set. We could have created the metric on the entire dataset. But, if we know that we're only interested in a portion of the talks, we should narrow our dataset before computing the metric.

In [6]:
#create the metric of the comments to views ratio
q_talks['comments_per_1000views']=1000*q_talks['comments']/q_talks['views']

#Sort talks in descending order of the ratio of views to comments
q_talks = q_talks.sort_values('comments_per_1000views', ascending=False)

#Print the top 10 talks
q_talks[['description', 'main_speaker', 'comments_per_1000views']].head(10)


description  \
803   David Bismark demos a new system for voting th...   
96    Richard Dawkins urges all atheists to openly s...   
694   Filmmaker Sharmeen Obaid-Chinoy takes on a ter...   
954   Janet Echelman found her true voice as an arti...   
840   Lesley Hazleton sat down one day to read the K...   
1787  Our consciousness is a fundamental aspect of o...   
661   Vaccine-autism claims, "Frankenfood" bans, the...   
443   At TED2009, Al Gore presents updated slides fr...   
732   In "THE 99," Naif Al-Mutawa's new generation o...   
1148  What does environmental devastation actually l...   

               main_speaker  comments_per_1000views  
803           David Bismark                1.534355  
96          Richard Dawkins                1.463841  
694   Sharmeen Obaid-Chinoy                1.420683  
954          Janet Echelman                1.359572  
840         Lesley Hazleton                1.285149  
1787         David Chalmers                1.235918  
661         Michael Specter                1.235704  
443                 Al Gore                1.091008  
732          Naif Al-Mutawa                0.981252  
1148             Garth Lenz                0.977574

# *Self-Assessment: Dealing with Dates - Solution*
This is straight out of the book. Apply is a handy function available in pandas that lets you run a function for each row or column of your data. You're seeing examples here of using a lambda (inline) function as well as using a separately created function (convert_int). 

The lambda function is just grabbing the year from the published date. It's doing that by splitting the string on the '-' character. This creates an array. We grab the first item in the array, which, if we had a valid date, should be the year. If we didn't have a valid date, then we drop in the np.nan.

In [7]:
#Convert release_date into pandas datetime format
ted['published_date'] = pd.to_datetime(ted['published_date'],
                                       errors='coerce', unit='s')

#see what the new date looks like
print("This is what the datetime string looks like:")
display(ted['published_date'].head())


#Extract year from the datetime
ted['published_year'] = ted['published_date'].apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

#Helper function to convert NaT to 0 and all other years to integers.
def convert_int(x):
    try:
        return int(x)
    except:
        return 0

#Apply convert_int to the year feature
ted['published_year'] = ted['published_year'].apply(convert_int)
    
ted.head()

This is what the datetime string looks like:


0   2006-06-27 00:11:00
1   2006-06-27 00:11:00
2   2006-06-27 00:11:00
3   2006-06-27 00:11:00
4   2006-06-27 20:38:00
Name: published_date, dtype: datetime64[ns]

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  \
0      Ken Robinson: Do schools kill creativity?            1   
1           Al Gore: Averting the climate crisis            1   
2                  David Pogue: Simplicity sells            1   
3             Majora Carter: Greening the ghetto            1   
4  Hans Rosling: The best stats you've ever seen            1   

       published_date                                            ratings  \
0 2006-06-27 00:11:00  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1 2006-06-27 00:11:00  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2 2006-06-27 00:11:00  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3 2006-06-27 00:11:00  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4 2006-06-27 20:38:00  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  published_year  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110            2006  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520            2006  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292            2006  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550            2006  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869            2006

# *Self-Assessment: Stringified Dictionaries - Solution*

This is also straight from the book. When we use the literal_eval function on the ratings column, we get a dictionary that we can manipulate. The "name" key holds the part of the ratings that we care about. We want to convert these words to lower case and create a list of the words. We create an empty list if there were no ratings.


In [8]:
#Import the literal_eval function from ast
from ast import literal_eval

#Convert all NaN into stringified empty lists
ted['ratings'] = ted['ratings'].fillna('[]')

#Apply literal_eval to convert stringified empty lists to the list object
ted['ratings'] = ted['ratings'].apply(literal_eval)


#Convert list of dictionaries to a list of strings
ted['ratings'] = ted['ratings'].apply(lambda x: [i['name'].lower() for i in x] if isinstance(x, list) else [])

#See how 'ratings' has changed?
ted.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  \
0      Ken Robinson: Do schools kill creativity?            1   
1           Al Gore: Averting the climate crisis            1   
2                  David Pogue: Simplicity sells            1   
3             Majora Carter: Greening the ghetto            1   
4  Hans Rosling: The best stats you've ever seen            1   

       published_date                                            ratings  \
0 2006-06-27 00:11:00  [funny, beautiful, ingenious, courageous, long...   
1 2006-06-27 00:11:00  [funny, courageous, confusing, beautiful, unco...   
2 2006-06-27 00:11:00  [funny, courageous, ingenious, beautiful, unco...   
3 2006-06-27 00:11:00  [courageous, beautiful, confusing, funny, inge...   
4 2006-06-27 20:38:00  [ingenious, funny, beautiful, courageous, long...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  published_year  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110            2006  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520            2006  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292            2006  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550            2006  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869            2006

# *Self-Assessment: Create the Knowledge-Based Recommender - Solution*

We're creating this as a function that takes in the dataframe and the percentile of views that we want to return. 
We'll first generate our list of unique words to present to users. We'll also stringify our list of ratings so we can use str.contains to filter.

In [9]:
def build_chart(gen_ted, percentile=0.1):
    #Show user the list of word ratings to choose from
    s = ', '.join(gen_ted.apply(lambda x:pd.Series(x['ratings']),axis=1).stack().unique())
    print(s)
    
    
    #Ask for preferred word rating
    print("Select a descriptive word from the list above for the 'word rating'")
    rating = input()
    
    #Ask for lower limit of film year
    print("Input earliest year published (2006 to 2017)")
    low_year = int(input())
    
    #Ask for upper limit of film year
    print("Input latest year published(2006 to 2017)")
    high_year = int(input())
    
    
    #Define a new talks variable to store the preferred talks. 
    #Copy the contents of gen_ted to talks
    talks = gen_ted.copy()
    
    #Filter based on the condition
    talks = talks[(talks['ratings'].apply(lambda x: rating in x)) & 
                    (talks['published_year'] >= low_year) & 
                    (talks['published_year'] <= high_year)]
    
    #Calculate the number of views for the  percentile 
    m = talks['views'].quantile(percentile)

    #Only consider movies that have higher than m votes. Save this in a new dataframe q_movies (note using .loc here prevents a warning)
    q_talks = talks.copy().loc[talks['views'] >= m]
    
    #create the metric of the comments to views ratio
    q_talks['comments_per_1000views']=1000*q_talks['comments']/q_talks['views']

    #Sort talks in descending order of the ratio of views to comments
    q_talks = q_talks.sort_values('comments_per_1000views', ascending=False)
    
    return q_talks

In [10]:
#Generate the chart for top talks for these user preferences and display top 5.
#Show the results for the word rating "obnoxious" and published years between 2009 and 2014.
gen_ted_final = build_chart(ted).head(5)

gen_ted_final[['main_speaker','name','published_year','comments_per_1000views']]

funny, beautiful, ingenious, courageous, longwinded, confusing, informative, fascinating, unconvincing, persuasive, jaw-dropping, ok, obnoxious, inspiring
Select a descriptive word from the list above for the 'word rating'


 ingenious

Input earliest year published (2006 to 2017)


Input latest year published(2006 to 2017)


 2017

main_speaker  \
694   Sharmeen Obaid-Chinoy   
954          Janet Echelman   
840         Lesley Hazleton   
1787         David Chalmers   
661         Michael Specter   

                                                   name  published_year  \
694   Sharmeen Obaid-Chinoy: Inside a school for sui...            2010   
954        Janet Echelman: Taking imagination seriously            2011   
840               Lesley Hazleton: On reading the Koran            2011   
1787  David Chalmers: How do you explain consciousness?            2014   
661       Michael Specter: The danger of science denial            2010   

      comments_per_1000views  
694                 1.420683  
954                 1.359572  
840                 1.285149  
1787                1.235918  
661                 1.235704

# *Self-Assessment: TF-IDF Vectors - Solution*
This is all straight from the book. More information about the TfidfVectorizer is available online here: https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html

In [11]:
#Import TfIdfVectorizer from the scikit-learn library
from sklearn.feature_extraction.text import TfidfVectorizer

#Define a TF-IDF Vectorizer Object. Remove all english stopwords
tfidf = TfidfVectorizer(stop_words='english')

#Replace NaN with an empty string
ted['description'] = ted['description'].fillna('')

#Construct the required TF-IDF matrix by applying the fit_transform method on the description feature
tfidf_matrix = tfidf.fit_transform(ted['description'])

#Output the shape of tfidf_matrix (rows first, then columns)
tfidf_matrix.shape

(2550, 15162)

In [12]:
#bonus - take a look some of the individual words in the description
feature_names = tfidf.get_feature_names()
feature_names[500:510]

['albatrosses',
 'albert',
 'alberta',
 'alberto',
 'albinism',
 'albright',
 'album',
 'albuquerque',
 'alchemists',
 'alcohol']

In [13]:
#bonus - this is saying that for the first document, none of the 500th to 510th words shown above show up in that document
tfidf_list = tfidf_matrix.toarray()
tfidf_list[0, 500:510]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

# *Self-Assessment: Create the Content-Based Recommender Based on Cosine Similarity - Solution*
This is also straight from the book. We don't expect you to understand everything to do with linear kernels. But if you're interested, the documentation is here:
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.linear_kernel.html



In [14]:
# Import linear_kernel to compute the dot product
from sklearn.metrics.pairwise import linear_kernel

# Compute the cosine similarity matrix
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

#Construct a reverse mapping of indices and talk names, and drop duplicate names, if any
indices = pd.Series(ted.index, index=ted['name']).drop_duplicates()

In [15]:
# Function that takes in talk title as input and gives recommendations 
def content_recommender(df, title, cosine_sim, indices, topN=2): 
    # Obtain the index of the movie that matches the title
    idx = indices[title]
    # Get the pairwsie similarity scores of all movies with that movie and convert to tuples
    sim_scores = list(enumerate(cosine_sim[idx]))
    #delete the movie that was passed in
    del sim_scores[idx]
    
    # Sort the movies based on the cosine similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    
    # Get the scores of the top-n most similar movies.
    sim_scores = sim_scores[:topN]
    
    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]
    
    # Return the top 10 most similar movies
    return df.iloc[movie_indices]


In [16]:
#Get recommendations for Tyler Cowen: Be suspicious of simple stories
content_recommender(ted, 'Tyler Cowen: Be suspicious of simple stories', cosine_sim, indices, 10)

comments                                        description  duration  \
1910        54  Making toast doesn’t sound very complicated --...       541   
451        324  Behavioral economist Dan Ariely, the author of...      1046   
1115       110  Cybercrime expert Mikko Hypponen talks us thro...       563   
1405       810  When war between Israel and Iran seemed immine...       897   
2361        31  Stories are necessary, but they're not as magi...       766   
406        280  Behavioral economist Dan Ariely studies the bu...       983   
775        229  Julian Treasure says our increasingly noisy wo...       434   
335         45  Would you negotiate with someone you knew to b...      1389   
145         83  The MIT Media Lab's John Maeda lives at the in...       959   
757        215  His Holiness the Karmapa talks about how he wa...      1523   

               event   film_date  languages              main_speaker  \
1910  TEDGlobal 2013  1370822400         28                 Tom Wujec   
451          EG 2008  1229040000         45                Dan Ariely   
1115    TEDxBrussels  1321488000         36            Mikko Hypponen   
1405  TEDxJaffa 2012  1346889600         33                Ronny Edry   
2361   TEDWomen 2016  1477526400         20           Sisonke Msimang   
406          TED2009  1233964800         36                Dan Ariely   
775   TEDGlobal 2010  1278633600         35           Julian Treasure   
335          TED2008  1201910400         20            Samantha Power   
145          TED2007  1172880000         24                John Maeda   
757    TEDIndia 2009  1257292800         29  His Holiness the Karmapa   

                                                   name  num_speaker  \
1910  Tom Wujec: Got a wicked problem? First, tell m...            1   
451   Dan Ariely: Are we in control of our own decis...            1   
1115       Mikko Hypponen: Three types of online attack            1   
1405         Ronny Edry: Israel and Iran: A love story?            1   
2361   Sisonke Msimang: If a story moves you, act on it            1   
406                    Dan Ariely: Our buggy moral code            1   
775       Julian Treasure: Shh! Sound health in 8 steps            1   
335   Samantha Power: A complicated hero in the war ...            1   
145                John Maeda: Designing for simplicity            1   
757   His Holiness the Karmapa: The technology of th...            1   

          published_date                                            ratings  \
1910 2015-02-05 16:01:26  [ingenious, fascinating, jaw-dropping, funny, ...   
451  2009-05-19 01:00:00  [ingenious, funny, informative, courageous, fa...   
1115 2012-01-18 14:54:24  [persuasive, informative, jaw-dropping, inspir...   
1405 2012-12-21 16:08:02  [beautiful, jaw-dropping, courageous, inspirin...   
2361 2017-01-13 16:19:34  [ingenious, inspiring, fascinating, informativ...   
406  2009-03-17 01:00:00  [ingenious, persuasive, fascinating, informati...   
775  2010-09-23 13:55:00  [unconvincing, obnoxious, fascinating, persuas...   
335  2008-11-11 01:00:00  [courageous, inspiring, persuasive, jaw-droppi...   
145  2007-09-20 08:38:00  [funny, beautiful, fascinating, inspiring, inf...   
757  2010-09-01 08:48:00  [confusing, longwinded, obnoxious, inspiring, ...   

                                          related_talks  \
1910  [{'id': 145, 'hero': 'https://pe.tedcdn.com/im...   
451   [{'id': 487, 'hero': 'https://pe.tedcdn.com/im...   
1115  [{'id': 1192, 'hero': 'https://pe.tedcdn.com/i...   
1405  [{'id': 1214, 'hero': 'https://pe.tedcdn.com/i...   
2361  [{'id': 462, 'hero': 'https://pe.tedcdn.com/im...   
406   [{'id': 420, 'hero': 'https://pe.tedcdn.com/im...   
775   [{'id': 660, 'hero': 'https://pe.tedcdn.com/im...   
335   [{'id': 272, 'hero': 'https://pe.tedcdn.com/im...   
145   [{'id': 2, 'hero': 'https://pe.tedcdn.com/imag...   
757   [{'id': 71, 'hero': 'https://pe.tedcdn.com/ima...   

                   